design-proposals/template-generation.md

```diff
@@ -0,0 +1,131 @@
+* **VMNamespace**: the namespace of the reference VM. If not specified the VM will be looked up in the namepace in which VirtualMachineTemplateRequest was created (optional field).
+* **TemplateName**: the name of the template object that will be generated, if omitted it will be auto-generated (otional field).
+* **Format**: the format of the output template (required field).
+
+The first implemented format will be the [OpenShift Template(https://docs.openshift.com/container-platform/4.9/openshift_images/using-templates.html#templates-writing_using-templates)] format.
+
+In order to support multiple formats, the controller will have to be extended to generate the correct output for those formats.
+
+## API Examples
+
+Given the following VM
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: vm-ephemeral
+spec:
+  template:
+    domain:
+      devices:
+        disks:
+        - disk:
+            bus: virtio
+          name: containerdisk
+      resources:
+        requests:
+          memory: 128Mi
+    volumes:
+    - containerDisk:
+        image: registry:5000/kubevirt/cirros-container-disk-demo:devel
+      name: containerdisk
+```
+
+After posting this VirtualMachineTemplateRequest object
+
+```yaml
+apiVersion: ssp.kubevirt.io/v1alpha1
+kind: VirtualMachineTemplateRequest
+metadata:
+  name: virtualmachinetemplaterequest-sample
+spec:
+  vmName: vmi-ephemeral
+  templateName: vmi-ephemeral-template
+  format: template.openshift.io/v1
+```
+
+The following template will be created
+
+```yaml
+apiVersion: v1
+kind: Template
+metadata:
+  name: vmi-ephemeral-template
+  labels: ${LABELS}
+objects:
+- apiVersion: kubevirt.io/v1
+  kind: VirtualMachine
+  metadata:
+    generateName: vmi-ephemeral
+  template:
+    spec:
+      domain:
+        devices:
+          disks:
+          - disk:
+              bus: ${DISK0_BUS}
+            name: containerdisk
+        resources:
+          requests:
+            memory: ${REQUESTS_MEMORY}
+      volumes:
+      - containerDisk:
+          image: ${VOLUME0_IMAGE}
+        name: containerdisk
+parameters:
+- description: disk0-bus
+  name: DISK0_BUS
+  value: virtio
+- description: requests-memory
+  name: REQUESTS_MEMORY
+  value: 128Mi
+- description: volume0-image
+  name: VOLUME0_IMAGE
+  value: registry:5000/kubevirt/cirros-container-disk-demo:devel
+```
+
+## Template generation
+
+The template generation mechanism follows some basic rules:
+
+* every leaf field (but name fields) in the `VirtualMachineInstanceSpec` will be parameterized and given a default value according to the reference VM which has been created from.
+* the VM name will be converted into a `generateName` field.
+* the `dataVolumeTemplates` field inside a `VirtualMachineSpec` will be preserved as is (as it's already templated).
+* the `LABELS` parameter will always be created, its content will be merged with the existing labels of the `VirtualMachine` object.
```

Comment:
What about `annotations`?

Reply:
OpenShift templates don't have a way (as far as I know) to do the same for annotations, that's why I didn't include them, you may argue that this design is supposed to be generic and it is but since the first format to be implemented will be OCP I've not added it as a tradeoff in favor of simplicity.

Reply:
I think ruling out annotation can affect functionality. See 
```
// This annotation is used to inject ignition data
	// Used on VirtualMachineInstance.
	IgnitionAnnotation           string = "kubevirt.io/ignitiondata"

`// This annotation is to keep virt launcher container alive when an VMI encounters a failure for debugging purpose
	KeepLauncherAfterFailureAnnotation string = "kubevirt.io/keep-launcher-alive-after-failure"`
```